cmd/verify.rb

```diff
@@ -1,79 +1,137 @@
+              bottle = formula.bottle_for_tag(bottle_tag)
+
+              if bottle.nil?
+                opoo "Bottle for tag #{bottle_tag.to_sym.inspect} is unavailable."
+                next
+              end
+              bottle.fetch
+              result = check_attestation(bottle, "Homebrew/homebrew-core")
+              if result[:verified]
+                if args.json?
+                  json_results.push(result[:data])
+                else
+                  ohai "Verified #{bottle.name} with tag #{bottle_tag}."
+                end
+              else
+                case result[:error]
+                when JSON::ParserError
+                  opoo "#{bottle.name} with tag #{bottle_tag} returned invalid json: #{result[:message]}"
+                  next
+                when COMMAND_ERROR
+                  opoo "#{bottle.name} with tag #{bottle_tag} unverified, checking backfill signature."
+                  backup_result = check_attestation(bottle, "trailofbits/homebrew-brew-verify")
+                  if backup_result[:verified]
+                    timestamp = backup_result.dig(:data, 0, "verificationResult", "verifiedTimestamps", 0,
+                                                  "timestamp")
+                    if timestamp.nil?
+                      opoo "Unable to verify #{bottle.name} with tag #{bottle_tag} " \
+                           "because json does not have signature timestamp in expected location."
+                      next
+                    end
+                    parsed_timestamp = DateTime.parse(timestamp)
+                    # Backfilled signatures all occured before this date so if we find
+                    # a backfilled signature after this date, we should disregard it.
+                    last_signature_date = DateTime.new(2024, 3, 14)
+                    if parsed_timestamp < last_signature_date
+                      if args.json?
+                        json_results.push(backup_result[:data])
+                      else
+                        ohai "Verified #{bottle.name} with tag #{bottle_tag} using backfill signature."
+                      end
+                    else
+                      opoo "Unable to verify #{bottle.name} with tag #{bottle_tag}. " \
+                           "Backfilled signature dated after last backfill date."
+                    end
+                  end
+                end
+              end
+            end
           end
+        end
+        return unless args.json?
 
-          bottle.fetch
+        puts json_results.to_json
+      end
+
+      def check_attestation(bottle, signing_repo)
+        cmd = "gh attestation verify #{bottle.cached_download} -R #{signing_repo} --format json 2>/dev/null"
+        output = IO.popen(cmd, &:read)
+        exit_status = $CHILD_STATUS.exitstatus
+        if exit_status != 0
+          return { verified: false, error: COMMAND_ERROR, message: "Command failed with status #{exit_status}" }
```

Comment:
Flagging another thing for follow-up: returning a `Hash` like this will probably be hard to maintain/test, we should either declare a type or use exception control flow here (e.g. just expecting the caller to handle the invalid error states as different exceptions).

Reply:
Added some todo's to make sure I remember.